redis-py patching support

ikreymer opened the pull request:
An attempt to support a subset of the redis-py api via monkey-patching, as requested in #4 

redis-py not added as dependency, error printed if its not installed.

Some basic tests included in test/patch.py.

ikreymer committed:
support for using rlite with redis-py via monkey-patching, fixes #4
patch_connection() patches redis.connection.Connection and redis.connection.ConnectionPool
tests added in test/patch.py

seppo0010 commented on hirlite/patch_conn.py:
```diff
@@ -0,0 +1,120 @@
+
+try:
+    from queue import Queue
+except:
```
`except ImportError:`

seppo0010 commented on hirlite/patch_conn.py:
```diff
@@ -0,0 +1,120 @@
+        return True
+
+    def send_command(self, *args):
+        #print(args)
```
Clean up all `print`, commented and not

seppo0010 commented on hirlite/patch_conn.py:
```diff
@@ -0,0 +1,120 @@
+    """
+
+    if no_redis:
+        print("redis package not found, please install redis-py via 'pip install redis'")
```
raise an Exception instead?

seppo0010 commented on test/patch.py:
```diff
@@ -0,0 +1,68 @@
+
+        self.assertEquals(set(['zset', 'hset']), set(res))
+
+
```
Remove empty lines at the end

seppo0010 commented on hirlite/patch_conn.py:
```diff
@@ -0,0 +1,120 @@
+
+    redis.connection.ConnectionPool = RliteConnectionPool
+    redis.client.ConnectionPool = RliteConnectionPool
+    redis.ConnectionPool = RliteConnectionPool
```
It'd be nice if this operation can be undone... I'm not sure if there's a common pattern, but maybe return a function that roll backs the changes?

Something like:

```
def patch_connection(filename=':memory:'):
    RedisConnection = redis.connection.Connection
    redis.connection.Connection = RliteConnection
    def unpatch():
        redis.connection.Connection = RedisConnection
    return unpatch
```

Actually, it may be better to support `__enter__` and `__exit__` as well, so people can do

```
from hirlite import patch_connection

with patch_connection():
    r = redis.StrictRedis()
```

and then there are no global side-effects outside the desired scope

seppo0010 commented on test/patch.py:
```diff
@@ -0,0 +1,68 @@
+
+class PatchConnTest(TestCase):
+    def setUp(self):
+        hirlite.patch_connection()
```
If you follow my previous comment about returning a function to undo the monkeypatch, please call it on tearDown and assert the proper Redis classes are back

seppo0010 commented on test/__init__.py:
```diff
@@ -11,4 +11,11 @@ def tests():
+
+    try:
+        from test import patch
+        suite.addTest(unittest.makeSuite(patch.PatchConnTest))
```
I don't think the tests you added run on Travis[1], I believe redis package is not installed... it'd be nice to have a build matrix so we can test with and without redis[2]

[1] https://travis-ci.org/seppo0010/rlite-py/jobs/186281584

[2] https://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix

ikreymer committed:
- add unpatch_connection() to undo patching, tests
- add patch() context manager version
- add travis env for testing with and without redis, add python 3.5
- cleanup remove unused print, blank lines, exception name

ikreymer commented:
Added the changes. Can also rebase off head if you'd like.

seppo0010 commented:
It looks good to me. I'm ready to merge it and publish a new version. Do you wanna make any change before that, or do you think it is ready?

ikreymer commented:
yeah, i think it's ready. Thanks!
